src/lib.rs

```diff
@@ -1069,6 +1089,50 @@ impl GooseAttack {
         (Some(logger_thread), Some(all_threads_logger))
     }
 
+    // Helper to spawn a throttle thread if configured.
+    async fn setup_throttle(
+        &self,
+    ) -> (
+        // A channel used by GooseClients to throttle requests.
+        Option<mpsc::Sender<bool>>,
+        // A channel used by parent to tell throttle the load test is complete.
+        Option<mpsc::Sender<bool>>,
+    ) {
+        // If the throttle isn't configured, return immediately.
+        if self.configuration.throttle_requests.is_none() {
+            return (None, None);
+        }
+
+        // Unwrap is safe here as we exit early if the throttle isn't configured.
+        let throttle_requests = self.configuration.throttle_requests.unwrap();
+
+        // Create a bounded channel allowing single-sender multi-receiver to throttle
+        // GooseUser threads.
+        let (all_threads_throttle, throttle_receiver): (mpsc::Sender<bool>, mpsc::Receiver<bool>) =
+            mpsc::channel(throttle_requests);
+
+        // Create a channel allowing the parent to inform the throttle thread when the
+        // load test is finished. Even though we only send one message, we can't use a
+        // oneshot channel as we don't want to block waiting for a message.
+        let (parent_to_throttle_tx, throttle_rx) = mpsc::channel(1);
+
+        // Launch a new thread for throttling, no need to rejoin it.
+        let _ = Some(tokio::spawn(throttle::throttle_main(
+            throttle_requests,
+            throttle_receiver,
+            throttle_rx,
+        )));
+
+        let mut sender = all_threads_throttle.clone();
+        // Fill all but one slot in the channel to avoid a burst of traffic during
+        // startup.
```

Comment:
It might be good to add somewhere as a comment in the helper that this implements leaky-bucket :)

Reply:
Unless I'm misunderstanding, it's not _really_ a leaky bucket. A leaky bucket would be if we're not blocking when the bucket fills up, instead leaking the overflow. Hence why I've been referring to it as a "modified leaky-bucket" as we the only reason this works is we block when the bucket fills, and _don't_ leak.

Reply:
Reading a little further, I see this is an implementation of the leaky bucket as a queue -- updating comment.

Reply:
```rust
        // We start from 1 instead of 0 to intentionally fill all but one slot in the
        // channel to avoid a burst of traffic during startup. The channel then provides
        // an implementation of the leaky bucket algorithm as a queue. Requests have to
        // add a token to the bucket before making a request, and are blocked until this
        // throttle thread "leaks out" a token thereby creating space. More information
        // can be found at: https://en.wikipedia.org/wiki/Leaky_bucket
```

Reply:
fixed